Update BlockPreview component to accept multiple Blocks to preview

marekhrabe opened the pull request:
## Description
`BlockPreview` currently only accepts a single Block. This PR updates this to allow it to (optionally) accept an array of blocks. It also simplifies the API by requiring a Block object rather than expecting Block properties as individual args. 

**Note:** the Reusable Blocks previews are not currently working due to this unrelated bug https://github.com/WordPress/gutenberg/issues/16808

## How has this been tested?
Firstly confirm existing behaviour has been preserved with the change in API. To do this, manually test all instances where `BlockPreview` is used and ensure all function as per `master` branch:
- Insert Quote block, open Block Switcher, see Previews render
- Open Block Inserter, scroll to Reusable Blocks, hover to see a preview (please see note above)

To test multiple Blocks being rendered you can:
- [Apply](https://stackoverflow.com/questions/12320863/how-do-you-take-a-git-diff-file-and-apply-it-to-a-local-branch-that-is-a-copy-o) this diff: https://gist.github.com/obenland/761894e1a5be40091e41214756b6a323
(Applying the diff is necessary to make this PR's under-the-hood changes visible)
- Create a Post with a lot of Blocks in it.
- In a quote or button block, open the Block switcher to see all of the Blocks from the page appear in the `BlockPreviewContent` below the standard Block Style previews.

![test](https://user-images.githubusercontent.com/77539/62156685-d99f6b00-b2e1-11e9-8411-58569b4956e3.gif)

### Places to check for previews

- Undo the additional diff from instructions above and test with code in this branch:
  - Block Inserter - top left of the screen the "plus" button. In the inserter, hover a reusable block (you might need to create one first) and you should see a preview on right. As noted above, you will see an error there. If you see an error, that means previews still work the same way and this PR didn't break it. *Error is good here. Seeing nothing would be bad.*
  - Block Switcher - for Button or Quote block it shows "Block Styles" as described above. Make sure both the thumbnail and the detail after hovering the thumbnail contain a preview of the style.

## Types of changes
New feature (non-breaking change which adds functionality)

## To Demo
Show multiple blocks being previewed.

marekhrabe committed:
export BlockPreview

youknowriad reviewed (commented):
Right now that component expects name, attributes and innerBlocks prop to render the preview. I think if we want to expose it, it would be better to allow passing `blocks` as an array instead of these granular props. What do you think?

marekhrabe commented:
It definitely sounds like a nicer API. Do you think we should update all current usages to use `blocks` or just add `blocks` as another way of using this component?

youknowriad commented:
> Do you think we should update all current usages to use blocks

I'd opt for this. It seems straightforward?

youknowriad commented:
Also, I think we might want to expose `BlockPreviewContent` using the `BlockPreview` name instead. It seems that the current `BlockPreview` adds an unnecessary "title" that might not be useful for all usage. Thoughts?

This might require some adjustements to classnames and things like that.

getdave commented:
Let's keep this in sync with https://github.com/WordPress/gutenberg/pull/16113

cc @jasmussen

marekhrabe commented:
I did some investigation during the weekend and I'm currently thinking between:

1) exporting `BlockPreview` (previously known as `BlockPreviewContent`), and `BlockPreviewFullSize`(previously BlockPreview)
2) replacing both usages with a single `BlockPreview` component which would have a prop that controls whether it's scaled (and how much) and I'd remove the "Preview" heading completely from our component, making sure it gets moved to places where `BlockPreview` is used from

The smaller size is used in "Styles" panel for example and the full size can be seen inside the block switcher or in block inserter for reusable blocks

getdave commented:
@marekhrabe I've found the two-component modal pretty confusing. I'd lean towards number 2 also.

marekhrabe commented:
I also liked the approach when doing some prototypes locally, it was just a bit more work than expected as those two components do very different things and all places they are used from have to be updated.

With the rate #16113 is getting worked on and introducing a lot of changes, I think it might be wise to wait with this PR for that one to be merged. Otherwise, we will be doing some duplicate work about scaling etc…

getdave committed:
Update to accept multiple Blocks

getdave committed:
Update reusable Blocks preview to use the new single blocks prop

marekhrabe commented on packages/block-editor/src/components/block-preview/index.js:
```diff
@@ -28,12 +28,22 @@ function BlockPreview( props ) {
+		return null;
+	}
+
+	const theBlocks = Array.isArray( blocks ) ? blocks : [ blocks ];
```
Let's use [`castArray` from lodash](https://lodash.com/docs/4.17.14#castArray). It's already being used in other places in Gutenberg that are able to accept a block or a list of blocks.

marekhrabe commented:
I'm working separately on the unification of `BlockPreview` & `BlockPreviewContent` (WIP #16801).

With @getdave, we plan to spend some more time on various issues around block previews. To be able to ship gradual updates to these components, I would suggest to keep this PR only about the move from `name, attributes, innerBlocks` into `blocks` and make sure all preview usages will be migrated to `blocks`.  **I'd suggest removing the `export` from this PR.**

Let's finish the support for `blocks` here, scaling in #16113 and write docs for the preview and export the component as the last thing. Would this make sense?

youknowriad commented on packages/block-editor/src/components/block-preview/index.js:
```diff
@@ -28,12 +28,22 @@ function BlockPreview( props ) {
+	const theBlocks = Array.isArray( blocks ) ? blocks : [ blocks ];
+
+	// Create new "clone" Blocks to show in the preview
+	const blocksToPreview = theBlocks.map( ( { name, attributes, innerBlocks } ) => {
```
Why do we need to clone the blocks? Can't we just rely on the caller? 
Also, if we do, there's a `cloneBlock` function somewhere I believe.

getdave replied:
I've re-tested and I don't think it is required. I feel like I introduced it cos there was an error. But looking more closely I think the result is the same either way.

getdave committed:
Remove unecessary clone of Blocks

Not sure why this was introduced.

getdave committed:
Remove export. This is being handled in another PR

See https://github.com/WordPress/gutenberg/pull/16801

getdave committed:
Simplify casting to array via lodash

youknowriad commented on packages/block-editor/src/components/block-styles/index.js:
```diff
@@ -125,12 +123,15 @@ function BlockStyles( {
-								} }
-								innerBlocks={ block.innerBlocks }
+								blocks={
+									{
```
The "cloneBlock" already takes care of merging attributes and also safely clones innerBlocks. It feels like it's well-suited for this usage here. (Same for the switcher and inserter menu)

getdave replied:
Thanks. That's much cleaner.

getdave committed:
Utilise cloneBlocks to safely merge attributes on blocks prop

getdave committed:
Spread reusable block initial attrs correctly

youknowriad commented on packages/block-editor/src/components/block-switcher/index.js:
```diff
@@ -161,9 +161,14 @@ export class BlockSwitcher extends Component {
-								name={ blocks[ 0 ].name }
-								attributes={ { ...blocks[ 0 ].attributes, className: hoveredClassName } }
-								innerBlocks={ blocks[ 0 ].innerBlocks }
+								blocks={ {
```
This could be a clone too.

getdave replied:
Fixed

youknowriad commented on packages/block-editor/src/components/inserter/menu.js:
```diff
@@ -351,7 +352,7 @@ export class InserterMenu extends Component {
 
 				{ hoveredItem && isReusableBlock( hoveredItem ) &&
-					<BlockPreview name={ hoveredItem.name } attributes={ hoveredItem.initialAttributes } />
+					<BlockPreview blocks={ cloneBlock( hoveredItem, ...hoveredItem.initialAttributes ) } />
```
Why the `...` ?

getdave replied:
Whoops! Yeh no idea. Fixed.

getdave committed:
Fix cloneBlock fn to check for innerBlocks before attempting to map

The `innerBlocks` of the `block` being cloned can be `undefined`. Therefore, by attempting to map these we trigger an error.

Fixed to introduce existence check before attempting to manipulate innerBlocks.

getdave committed:
Simplify modifying passed block attrs via cloneBlock

getdave committed:
Removes unecessary spread operation

initalAttributes is already an object so no need to spread into an object.

youknowriad commented on packages/block-editor/src/components/block-preview/index.js:
```diff
@@ -28,12 +32,17 @@ function BlockPreview( props ) {
+		return null;
+	}
+
+	const theBlocks = castArray( blocks ).map( ( { name, attributes, innerBlocks } ) => createBlock( name, attributes, innerBlocks ) );
```
Still not sure why we dod this and not pass the blocks directly?

getdave replied:
Yeh again, when I don't do this `createBlock` the editor crashes. This code is as per the `master` branch only I'm mapping over a collection of Blocks rather than a single Block.

I know that Reusable Blocks passes `block.initialAttributes` as `attributes` rather than `block.attributes` and so it could be that this is why we need to create new Blocks?

Needs testing to see what happens if we remove this line and just pass Blocks directly.

retrofox replied:
The issue here is we forgot to cast the `value` property, so when [the flattenBlocks() function tries to destruct it](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/store/reducer.js#L90) and rebuild it as the new `stack` array it triggers the Uncaught TypeError: object is not iterable (cannot read property error). Pretty easy to reproduce it:

```js
var obj = { a: 0, b: 1, c: 2, d: 3, e: 4 };
var arr = [ ...obj ];
```

Adding the commit to cast the `value` avoiding the creating blocks flow.

marekhrabe replied:
> I know that Reusable Blocks passes block.initialAttributes as attributes rather than block.attributes and so it could be that this is why we need to create new Blocks?

This is what pointed me into the right direction. Blocks only have `attributes`, never `initialAttributes`. We've been using a wrong object with `cloneBlock`! Longer explanation https://github.com/WordPress/gutenberg/pull/16033#discussion_r308998056

youknowriad commented on packages/blocks/src/api/factory.js:
```diff
@@ -91,15 +91,21 @@ export function createBlock( name, attributes = {}, innerBlocks = [] ) {
+	// Check for innerBlocks on block before attempting
+	// to map them. There are edge cases where they can
+	// be `undefined`
+	if ( ! newInnerBlocks && block.innerBlocks ) {
```
What are these edge cases?

youknowriad replied:
Could be a signal of a missing `createBlock` call or something like that?

getdave replied:
I need to trace the instances where `blocks.innerBlocks` was `undefined`. All my calls are using `createBlock`/`cloneBlock` so I don't _think_ it was that but...

retrofox replied:
I think it worths doing this checking, maybe just changing/updating the comment?

marekhrabe replied:
I think I've solved the mystery! `innerBlocks` were truly undefined for all previews of reusable blocks called from `inserter/menu.js` and it caused bugs described above (and in one more review thread in this PR). 

This has happened because we've called `cloneBlocks` with the "block" object and new attributes. However, that "block" object was an **object describing block**, but not a block itself. This is the object we've been using:

https://github.com/WordPress/gutenberg/blob/823ba362fa3ca0016ef5f690e8f267420b994697/packages/block-editor/src/store/selectors.js#L1233-L1245

It shares some properties with a true block object (one created by `createBlock`) - `id` and `name` which made it kinda work with `cloneBlock` but it is definitely missing `innerBlocks` for example.

Knowing now that we don't actually have any block ready for a preview in the inserter, I have replaced the `cloneBlock` with a `createBlock` and that solved all issues around `innerBlocks`.

> Could be a signal of a missing createBlock call or something like that?

Spot on but it was a journey to discover why exactly 😀

marekhrabe replied:
Fixed in https://github.com/WordPress/gutenberg/pull/16033/commits/2b8f1fb6527278eff3c6e0dcac78d28cbbbd7619

getdave replied:
Nice sleuthing. This is what I had in mind to check this morning so thanks for sorting it.

retrofox committed:
block-preview: ensuring to cast BlockEditorProvider value prop as an Array

retrofox commented:
I've been testing the preview for the components that have defined the styles.

**master branch**
![image](https://user-images.githubusercontent.com/77539/62161660-145ad080-b2ed-11e9-9bcf-88b4631a1a59.png)

**this branch**

![image](https://user-images.githubusercontent.com/77539/62161963-9f3bcb00-b2ed-11e9-9f09-e712afd80cbd.png)

Although showing the `+` button into the preview already happens in prod, I wonder why it looks darker than in the master branch

marekhrabe commented:
I've done a rebase against the master branch (clean, no conflicts) so I can start integrating with #16801.

marekhrabe committed:
don't call cloneBlock on a non-block object

marekhrabe committed:
bring back the old behavior of cloneBlock

marekhrabe commented:
I have fixed the issue around `innerBlocks` being undefined which caused an avalanche of bugs in random places. https://github.com/WordPress/gutenberg/pull/16033#discussion_r308998056

getdave commented:
> I've been testing the preview for the components that have defined the styles.
> 
> **master branch**
> ![image](https://user-images.githubusercontent.com/77539/62161660-145ad080-b2ed-11e9-9bcf-88b4631a1a59.png)
> 
> **this branch**
> 
> ![image](https://user-images.githubusercontent.com/77539/62161963-9f3bcb00-b2ed-11e9-9f09-e712afd80cbd.png)
> 
> Although showing the `+` button into the preview already happens in prod, I wonder why it looks darker than in the master branch

Suggest we filter this out if previews along with any other Block inserters.

youknowriad reviewed (approved):
LGTM 👍Thanks for the updates. 

I agree that the inserter on the left should be hidden on previews. for me it shows similarly (color) in both master and this branch.

retrofox commented on packages/block-editor/src/components/block-preview/index.js:
```diff
@@ -1,8 +1,12 @@
 /**
  * WordPress dependencies
  */
 import { __ } from '@wordpress/i18n';
```
need to remove this function
